### PR TITLE
Reimplement split shard

### LINKED ISSUE #248: Reimplement split shard
Using the A -> B + C approached discussed in the last week.

### PR DESCRIPTION
**Pull request by zhangxu19830126**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [x] Code Refactoring

**Which issue(s) this PR fixes:**

Fixes #248 

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: raftstore/store_bootstrap.go [hunk 1]
```diff
@@ -198,18 +198,25 @@ func (s *store) doCreateInitShard(shard *Shard) {
...
 		var ids []uint64
-		for _, s := range v {
-			ids = append(ids, s.ID)
-			sm = append(sm, storage.ShardMetadata{
-				ShardID:  s.ID,
+		for _, shard := range v {
+			ids = append(ids, shard.ID)
+			sm = append(sm, meta.ShardMetadata{
+				ShardID:  shard.ID,
 				LogIndex: 0,
```
Comment (zhangxu19830126) [inline raftstore/store_bootstrap.go:207]: LogIndex need set to 1

File: raftstore/replica.go [hunk 2]
```diff
@@ -472,5 +478,59 @@ func getRaftConfig(id, appliedIndex uint64, lr *LogReader, cfg *config.Config) *
...
+
+	if wc == nil {
+		wc = ldb.NewWorkerContext()
 	}
+	return ldb.SaveRaftState(state.Shard.ID,
+		replica.ID,
+		raft.Ready{
+			Entries: []raftpb.Entry{{Index: 1, Term: 1, Type: raftpb.EntryNormal,
+				Data: protoc.MustMarshal(&rb)}},
+			HardState: raftpb.HardState{Commit: 1, Term: 1, Vote: replica.ID},
```
Comment (zhangxu19830126) [inline raftstore/replica.go:533]: ignore vote

File: raftstore/store.go [hunk 3]
```diff
@@ -125,7 +125,13 @@ func NewStore(cfg *config.Config) Store {
 		logdb:     logdb.NewKVLogDB(kv, logger.Named("logdb")),
 		stopper:   syncutil.NewStopper(),
 	}
+
 	s.vacuumCleaner = newVacuumCleaner(s.vacuum)
+	// TODO: make maxWaitToChecker configurable
+	s.splitChecker = newSplitChecker(4, uint64(s.cfg.Replication.ShardCapacityBytes),
```
Comment (zhangxu19830126) [inline raftstore/store.go:131]: move shardCapacityBytes to data storage

File: aware/shard.go [hunk 4]
```diff
@@ -24,7 +24,7 @@ type ShardStateAware interface {
 	// Splited the shard was splited on the current store
 	Splited(meta.Shard)
 	// Destory the shard was destoryed on the current store
-	Destory(meta.Shard)
+	Destoryed(meta.Shard)
```
Comment (zhangxu19830126) [inline aware/shard.go:27]: typo

File: raftstore/replica_split.go [hunk 5]
```diff
@@ -18,38 +18,63 @@ import (
...
+}
 
-	if err := pr.doCheckSplit(); err != nil {
-		pr.logger.Fatal("fail to add split check job",
-			zap.Error(err))
-	}
-	pr.sizeDiffHint = 0
+func (pr *replica) needDoCheckSplit() bool {
+	return pr.useCustomSplitCheck() ||
+		pr.sizeDiffHint >= uint64(pr.cfg.Replication.ShardSplitCheckBytes)
```
Comment (zhangxu19830126) [inline raftstore/replica_split.go:65]: rename sizeDiffHint

File: raftstore/split_checker.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	if sc.mu.running {
+		return
+	}
+
+	sc.mu.running = true
+	go func() {
+		for {
```
Comment (zhangxu19830126) [inline raftstore/split_checker.go:68]: instead of  stopper

File: raftstore/split_checker_test.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+	assert.False(t, sc.mu.running)
+	sc.start()
+	assert.True(t, sc.mu.running)
+
+	sc.close()
+	assert.False(t, sc.mu.running)
+}
+
+func TestSplitCheckerDoCheck(t *testing.T) {
```
Comment (zhangxu19830126) [inline raftstore/split_checker_test.go:49]: add comments

File: raftstore/replica_state_machine_exec.go [hunk 8]
```diff
@@ -155,108 +154,153 @@ func (d *stateMachine) doExecSplit(ctx *applyContext) (rpc.ResponseBatch, error)
...
+		d.logger.Fatal("invalid splits keys",
+			log.HexField("actual-start", splitReqs.Requests[0].Start),
+			log.HexField("shard-start", current.Start),
+			log.HexField("actual-end", splitReqs.Requests[len(splitReqs.Requests)-1].End),
+			log.HexField("shard-end", current.End))
+	}
+
+	sort.Slice(current.Replicas, func(i, j int) bool {
+		return current.Replicas[i].ID < current.Replicas[j].ID
+	})
```
Comment (zhangxu19830126) [inline raftstore/replica_state_machine_exec.go:172]: move to split checker

File: raftstore/replica_apply_result.go [hunk 9]
```diff
@@ -162,65 +161,62 @@ func (pr *replica) applyConfChange(cp *configChangeResult) {
...
+		var fields []zap.Field
+		fields = append(fields, log.ShardField("old", pr.getShard()))
+		for idx, s := range result.newShards {
+			fields = append(fields, log.ShardField(fmt.Sprintf("new-%d", idx), s))
+		}
+		ce.Write(fields...)
 	}
 
-	for _, shard := range result.shards {
+	estimatedSize := pr.approximateSize / uint64(len(result.newShards))
```
Comment (zhangxu19830126) [inline raftstore/replica_apply_result.go:176]: add some comments

File: raftstore/replica_apply_result.go [hunk 10]
```diff
@@ -241,6 +237,11 @@ func (pr *replica) applySplit(result *splitResult) {
 			log.ShardField("new-shard", shard))
 	}
 
+	// all new shards started, update key range atomic
+	pr.store.updateShardKeyRange(result.newShards[0].Group, result.newShards...)
+
+	// current shard not used, keep data.
+	pr.store.destroyReplica(pr.shardID, true, false, "splited")
```
Comment (zhangxu19830126) [inline raftstore/replica_apply_result.go:244]: add comments on false

**Review by zhangxu19830126 — COMMENTED** [review]
LGTM

**Review by lni — APPROVED** [review]
Reviewed together with @zhangxu19830126. 

LGTM

**Pull request merged**